### PR TITLE
bugfix: metrics client is initialized 2 times per shard

### PR DESCRIPTION
**Pull request by wxing1292**
solves #477

### CODE REVIEW

**Comment by vancexu** [comment]
Where is the 2 times initialization in old implementation?

**Comment by wxing1292** [comment]
@vancexu cassandraPersistenceClientFactory.go and shardContext.go

**Comment by wxing1292** [comment]
@yiminc 
https://github.com/uber/cadence/pull/484#pullrequestreview-85760683
A1: I change the # of shards from 4 to 16384, and launch the service, check the memory usage using pprof, which is ~7G (used by this tagged metrics client), I change to this bugfix, and redo the check, the memory usage is ~ 3-4 G. so issue fixed.

A2 & A3: since the tagged metrics client is per shard (from the original code), i assume that we need to continue to do that.

File: service/history/shardController_test.go [hunk 1]
```diff
@@ -94,7 +94,7 @@ func (s *shardControllerSuite) TestAcquireShardSuccess() {
 		if hostID == 0 {
 			myShards = append(myShards, shardID)
 			mockExecutionMgr := &mmocks.ExecutionManager{}
-			s.mockExecutionMgrFactory.On("CreateExecutionManager", mock.Anything).Return(mockExecutionMgr, nil).Once()
+			s.mockExecutionMgrFactory.On("CreateExecutionManager", mock.Anything, mock.Anything).Return(mockExecutionMgr, nil).Once()
```
Comment (yiminc) [inline service/history/shardController_test.go]: do you still need this change?

File: common/persistence/cassandraPersistenceClientFactory.go [hunk 2]
```diff
@@ -30,14 +30,14 @@ import (
 type (
 	cassandraPersistenceClientFactory struct {
 		session       *gocql.Session
-		metricsClient metrics.Client
 		logger        bark.Logger
+		metricsClient metrics.Client
```
Comment (yiminc) [inline common/persistence/cassandraPersistenceClientFactory.go]: revert this

File: service/history/service.go [hunk 3]
```diff
@@ -214,7 +214,8 @@ func (s *Service) Start() {
 		p.CassandraConfig.Keyspace,
 		s.config.ExecutionMgrNumConns,
 		p.Logger,
-		base.GetMetricsClient())
```
Comment (madhuravi) [inline service/history/service.go]: What does this method do? Should it be removed if no longer used?

Comment (wxing1292) [reply service/history/service.go]: this is actually referenced at line 138:
https://github.com/uber/cadence/pull/484/files/b41eff2eeeca590eba88c5cb7ef67de7977ac034#diff-48531afaf74492193b35d730980b6ecdR138

Comment (wxing1292) [reply service/history/service.go]: @madhuravi 
this is actually referenced at line 138:
https://github.com/uber/cadence/pull/484/files/b41eff2eeeca590eba88c5cb7ef67de7977ac034#diff-48531afaf74492193b35d730980b6ecdR138

File: common/persistence/cassandraPersistenceClientFactory.go (hunk 2 above)
Comment (wxing1292) [reply common/persistence/cassandraPersistenceClientFactory.go]: we need this field, so the factory can use the metrics client if provided

Comment (wxing1292) [reply common/persistence/cassandraPersistenceClientFactory.go]: fixed

File: service/history/shardController.go [hunk 4]
```diff
@@ -107,7 +107,7 @@ func newShardController(host *membership.HostInfo, resolver membership.ServiceRe
 
 func newHistoryShardsItem(shardID int, shardMgr persistence.ShardManager, historyMgr persistence.HistoryManager,
 	metadataMgr persistence.MetadataManager, executionMgrFactory persistence.ExecutionManagerFactory, factory EngineFactory,
-	host *membership.HostInfo, config *Config, logger bark.Logger, reporter metrics.Client) (*historyShardsItem, error) {
+	host *membership.HostInfo, config *Config, logger bark.Logger, metricsClient metrics.Client) (*historyShardsItem, error) {
```
Comment (yiminc) [inline service/history/shardController.go]: nit: revert this as well.

File: common/persistence/cassandraPersistenceClientFactory.go [hunk 5]
```diff
@@ -37,7 +37,7 @@ type (
 
 // NewCassandraPersistenceClientFactory is used to create an instance of ExecutionManagerFactory implementation
 func NewCassandraPersistenceClientFactory(hosts string, port int, user, password, dc string, keyspace string,
-	numConns int, logger bark.Logger, mClient metrics.Client) (ExecutionManagerFactory, error) {
+	numConns int, logger bark.Logger, metricsClient metrics.Client) (ExecutionManagerFactory, error) {
```
Comment (yiminc) [inline common/persistence/cassandraPersistenceClientFactory.go]: nit: revert this

**Pull request merged**